osc.js

```diff
@@ -1,48 +1,174 @@
 const { InstanceBase, Regex, runEntrypoint } = require('@companion-module/base')
 const OSCRawClient = require('./osc-raw.js');
 const OSCTCPClient = require('./osc-tcp.js');
-const UpgradeScripts = require('./upgrades')
+const OSCUDPClient = require('./osc-udp.js');
+const UpgradeScripts = require('./upgrades');
+const { resolveHostname, isValidIPAddress, parseArguments, evaluateComparison } = require('./helpers.js');
 
 class OSCInstance extends InstanceBase {
 	constructor(internal) {
 		super(internal)
 	}
 
+	//Initialization
 	async init(config) {
-		this.config = config
-		this.tcpClient = new OSCTCPClient(this, this.config.host, this.config.port, this.config.listen)
-		this.rawClient = new OSCRawClient(this, this.config.host, this.config.port, this.config.listen)
+		this.config = config;
+		this.targetHost;
+		this.udpClient;
+		this.tcpClient;
+		this.rawClient;
+		
+		this.onDataReceived = {};
+
+		let validate = false;
+
+		if (this.config.host) {
+			if (!isValidIPAddress(this.config.host)) {
+				await resolveHostname(this, this.config.host)
+				.then ((ip) => {
+					this.targetHost = ip;
+					validate = true;
+				})
+				.catch(err => {
+					this.log('error', `Unable to resolve hostname for ${this.config.host}: ${err.message}`);
+					this.updateStatus('bad_config');
+					validate = false;
+				});
+			} else {
+				this.targetHost = this.config.host;
+				validate = true;
+			}
+		}
+
+		if (this.config.listen) {
+			if (this.targetHost && (this.config.targetPort || this.config.feedbackPort)) {
 
-		this.updateStatus('ok')
+				if (this.config.feedbackPort) {
+					this.udpClient = new OSCUDPClient(this, this.targetHost, this.config.feedbackPort, this.config.listen);
+				}
+				
+				if (this.config.targetPort) {
+					this.tcpClient = new OSCTCPClient(this, this.targetHost, this.config.targetPort, this.config.listen);
+					this.rawClient = new OSCRawClient(this, this.targetHost, this.config.targetPort, this.config.listen);
+				}
+				
+				if (validate) {
+					this.setupListeners();
+				}
+				
+			}
+		} else {
+			this.updateStatus('ok');
+		}
 
-		this.updateActions() // export actions
+		this.updateActions(); // export actions
+		this.updateFeedbacks(); // export feedback
+		
 	}
 	// When module gets deleted
 	async destroy() {
 		this.log('debug', 'destroy')
 	}
-
+	  
 	async configUpdated(config) {
-		this.config = config
-		this.tcpClient = new OSCTCPClient(this, this.config.host, this.config.port)
-		this.rawClient = new OSCRawClient(this, this.config.host, this.config.port)
+		this.config = config;
+
+		if (this.udpClient && this.udpClient.isConnected()) {
```

Comment:
I know you didn't start this, but it looks like each of classes for these 3 client types have the same methods/api exposed, so presumably things could be simplified by having a single `this.client` property which is setup as the correct type? That would avoid if statements like this, as it will always be `this.client`

Again, it's not caused by this so I don't mind if you don't want to change it

Reply:
I did start this actually, since I authored those original additions for TCP and Raw sockets. I've refactored to condense it to a single this.client now.